src/Features/Core/Portable/GenerateComparisonOperators/GenerateComparisonOperatorsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,276 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeGeneration;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.GenerateComparisonOperators
+{
+    using static CodeGenerationSymbolFactory;
+
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, LanguageNames.VisualBasic), Shared]
+    internal class GenerateComparisonOperatorsCodeRefactoringProvider : CodeRefactoringProvider
+    {
+        private const string LeftName = "left";
+        private const string RightName = "right";
+
+        private static ImmutableArray<CodeGenerationOperatorKind> s_operatorKinds =
+            ImmutableArray.Create(
+                CodeGenerationOperatorKind.LessThan,
+                CodeGenerationOperatorKind.LessThanOrEqual,
+                CodeGenerationOperatorKind.GreaterThan,
+                CodeGenerationOperatorKind.GreaterThanOrEqual);
+
+        [ImportingConstructor]
+        public GenerateComparisonOperatorsCodeRefactoringProvider()
+        {
+        }
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var (document, textSpan, cancellationToken) = context;
+
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var sourceText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            // We offer the refactoring when the user is either on the header of a class/struct,
+            // or if they're between any members of a class/struct and are on a blank line.
+            if (!syntaxFacts.IsOnTypeHeader(root, textSpan.Start, fullHeader: true, out var typeDeclaration) &&
+                !syntaxFacts.IsBetweenTypeMembers(sourceText, root, textSpan.Start, out typeDeclaration))
```

Comment:
nit: Does this include lines that are braces only? 

```csharp
class C 
{ // <-- Like this line
}
```

Reply:
not sure. however, the core goal here is that this is the normal logic we use for a bunch of htese features so that they behave consistently.  If we wanted to support such a scenario (and this helper didn't do that), we would just update the helper.

Reply:
sounds good